Detectors/PHOS/simulation/src/Detector.cxx

```diff
@@ -0,0 +1,1029 @@
+    for(Int_t itr=mHits->size()-1; itr>=0; itr--){
+      Hit *h = &(mHits->at(itr)) ;
+      if(h->GetTrackID()!=superParent) //switched to another SuperParent, do not search further
+        break ;
+      if(h->GetDetectorID() == detID){ //found correct hit
+        h->AddEnergyLoss(lostenergy);
+        mCurentSuperParent=superParent ;
+        mCurrentTrackID = partID ;
+        mCurrentCellID = detID ;
+        mCurrentHit = h;
+        return true ;
+      }
+    }
+
+  }
+  //Create new Hit
+  Double_t posX, posY, posZ, momX, momY, momZ, energy;
+  mcapp->TrackPosition(posX, posY, posZ);
+  mcapp->TrackMomentum(momX, momY, momZ, energy);
+  Double_t estart = mcapp->Etot();
+  Double_t time = mcapp->TrackTime() * 1.e+9; // time in ns?? To be consistent with EMCAL
+   
+  mCurrentHit = AddHit(superParent, detID, Point3D<float>(float(posX), float(posY), float(posZ)),
+                       Vector3D<float>(float(momX), float(momY), float(momZ)),estart, time, lostenergy);
+  mCurentSuperParent=superParent ;
+  mCurrentTrackID = partID ;
+  mCurrentCellID = detID ;
+  return true;
+     
+}
+
+
+Hit* Detector::AddHit(Int_t trackID, Int_t detID,
+                      const Point3D<float>& pos, const Vector3D<float>& mom, Double_t totE, Double_t time, Double_t eLoss)
+{
+  LOG(DEBUG4) << "Adding hit for track " << trackID << " with position (" << pos.X() << ", "
+              << pos.Y() << ", " << pos.Z() << ") and momentum (" << mom.X() << ", " << mom.Y() << ", " << mom.Z()
+              << ")  with energy " << totE << " loosing " << eLoss << std::endl;
+  mHits->emplace_back(trackID, detID, pos, mom, totE, time, eLoss);
+  return &(mHits->back());
+}
+
+
+
+void Detector::ConstructGeometry()
+{
+  //Create geometry description of PHOS depector for Geant simulations.
+
+  using boost::algorithm::contains;
+  LOG(DEBUG) << "Creating PHOS geometry\n";
+
+  PHOS::GeometryParams * geom = PHOS::GeometryParams::GetInstance("Run2"); 
+
+  if (! geom) {
+    LOG(ERROR) << "ConstructGeometry: PHOS Geometry class has not been set up.\n";
+  }
+
+  //Configure geometry So far we have onny one: Run2
+  {
+      mCreateCPV=kTRUE ;
+      mCreateHalfMod=kTRUE;
+      mActiveModule[0]=kFALSE ;
+      mActiveModule[1]=kTRUE ;
+      mActiveModule[2]=kTRUE ;
+      mActiveModule[3]=kTRUE ;
+      mActiveModule[4]=kTRUE ;
+      mActiveModule[5]=kFALSE ;
+      mActiveCPV[0]=kFALSE ;
+      mActiveCPV[1]=kFALSE ;
+      mActiveCPV[2]=kFALSE ;
+      mActiveCPV[3]=kTRUE ;
+      mActiveCPV[4]=kFALSE ;
+      mActiveCPV[5]=kFALSE ;
+  }
+
+
+
+  //First create necessary materials
+  CreateMaterials();
+
+  //Create a PHOS modules-containers which will be filled with the stuff later.
+  //Depending on configuration we should prepare containers for normal PHOS module "PHOS"
+  // PHOS module with CPV in front "PHOSC" and half-module "PHOH"
+  TVirtualMC::GetMC()->Gsvolu("PHOS", "TRD1", getMediumID(ID_AIR), geom->GetPHOSParams(), 4) ;        
```

Comment:
please avoid `TVirtualMC::GetMC()`. The base Detector class already has a cached reference to this instance.